main/src/library/Graph.flix

```diff
@@ -101,4 +101,68 @@ namespace Graph {
         Reachable(n1, n2) :- Reachable(n1, m), Edge(m, n2).
     }
 
+    ///
+    /// Optionally returns the shortest distance from `src` to `dst` in the weighted directed graph `g`.
+    ///
+    pub def distance(src: {src :: t}, dst: {dst :: t}, g: m[(t, d, t)]): Option[d] with Foldable[m], Boxable[t], Boxable[d], Add[d] =
+        let edges = inject g into Edge;
+        let x = query edges, dists(src.src) select x from Distance(n, x) where n == dst.dst;
+        List.minimum(x)
+
+    ///
+    /// Returns the shortest distance from `src` to every other vertex in the weighted directed graph `g`.
+    ///
+    pub def distances(src: t, g: m[(t, d, t)]): Map[t, Option[d]] with Foldable[m], Boxable[t], Boxable[d], Add[d] =
+        let edges = inject g into Edge;
+        let allDists = #{
+            // Collects all the distances `d` from `src` to `n` into a Set.
+            AllDists(n; Set#{d}) :- Distance(n, d).
+        };
+        let reachableDists = query edges, dists(src), allDists select (x, Set.minimum(s)) from AllDists(x; s) |> List.toMap;
+        let unreachableDists = unreachableWithDists(src, g)
+                                |> List.foldRight((x, acc) -> Map.insert(x, None, acc), Map.empty());
+        Map.union(reachableDists, unreachableDists)
+
+    ///
+    /// Helper function for `distance` and `distances`.
+    /// Returns a Datalog program which computes distances to all reachable nodes when given a set of `Edge` facts and a `src`.
+    ///
+    def dists(src: t): #{ Edge(t, d, t) | r } with Boxable[t], Boxable[d], Add[d] = #{
+        // Distance `d` from `src` to `n`.
+        Distance(n, d) :- Edge(src, d, n), if n != src.
+        // If `src` can reach `m` with distance `d1` and
+        // there is an edge from `m` to `n` with distance `d2`
+        // then `src` can reach `n` with distance `d1 + d2`.
+        Distance(n, d1 + d2) :- Distance(m, d1), Edge(m, d2, n), if n != src.
```

Comment:
Aside from the lattice stuff noted above, you need to only consider the minimum distance since finding all distances is potentially infinite, as you've seen, and potentially slow even with looping removed. An example could be something like nodes A, B, and C, where there are ten edges from A to B and ten edges from B to C. With all distances, there are upwards of 110 distances instead of 3.